node/src/components/consensus/era_supervisor.rs

```diff
@@ -595,6 +630,96 @@ where
+        timestamp: Timestamp,
+        genesis_start_time: Timestamp,
+    ) -> Effects<Event<I>> {
+        let last_activation_point = self.era_supervisor.protocol_config.last_activation_point;
+        let mut effects: Effects<Event<I>> = Default::default();
+
+        for era_id in self
+            .era_supervisor
+            .iter_past(current_era, self.era_supervisor.bonded_eras * 2)
+        {
+            let maybe_switch_block = if era_id > last_activation_point {
+                Some(EraId(era_id.0 - 1))
+            } else {
+                None
+            }
+            .and_then(|switch_id| switch_blocks.get(&switch_id));
+
+            let newly_slashed = maybe_switch_block
+                .and_then(|bhdr| bhdr.era_end())
+                .map(|era_end| era_end.equivocators.clone())
+                .unwrap_or_default();
+
+            let slashed = self
+                .era_supervisor
+                .iter_past_other(era_id, self.era_supervisor.bonded_eras)
+                .filter_map(|old_id| switch_blocks.get(&old_id).and_then(|bhdr| bhdr.era_end()))
+                .flat_map(|era_end| era_end.equivocators.clone())
+                .collect();
+
+            let booking_era_id = EraId(
+                era_id
+                    .0
+                    .saturating_sub(self.era_supervisor.protocol_config.auction_delay + 1),
+            );
+
+            let seed = if let Some(booking_block) = switch_blocks.get(&booking_era_id) {
+                EraSupervisor::<I>::era_seed(booking_block.hash(), booking_block.accumulated_seed())
+            } else {
+                0
+            };
+
+            let (validators, start_height, state_root_hash) =
+                if let Some(switch_block) = maybe_switch_block {
+                    (
+                        switch_block
+                            .next_era_validator_weights()
+                            .cloned()
+                            .expect("switch block should have era validator weights"),
+                        switch_block.height() + 1,
+                        *switch_block.state_root_hash(),
+                    )
+                } else {
+                    // TODO: the start height should be 1 more than the last block before the last
+                    // upgrade, and the state hash should be the (potentially modified) state hash
+                    // from that block
+                    (validators.clone(), 0, genesis_state_root_hash)
+                };
+
+            let results = self.era_supervisor.new_era(
+                current_era,
+                timestamp,
+                validators,
+                newly_slashed,
+                slashed,
+                seed,
+                genesis_start_time, /* TODO: this should also be defined in some reasonable way
+                                     * based on either the switch block or chainspec / upgrade
+                                     * point */
```

Comment:
Can we add this to the
```rust
let (validators, start_height, state_root_hash) =
```
case distinction as well, and use the switch block's timestamp as the start time if applicable?